falcon/cluster/cluster.py

```diff
@@ -547,10 +701,392 @@ def _get_cluster_medoids(
+        bins_peaks.append(np.empty(0, np.float32))
+
+    for spec in spectra:
+        for mz, intensity in zip(spec.mz, spec.intensity):
+            bin_idx = math.floor((mz - min_mz) / bin_size)
+            if 0 <= bin_idx < n_bins:
+                if bins_indices[bin_idx] == -1:
+                    bins_indices[bin_idx] = bin_idx
+                bins_peaks[bin_idx] = np.append(bins_peaks[bin_idx], intensity)
+    # Remove empty bins
+    mask = bins_indices != -1
+    bins_indices = bins_indices[mask]
+    bins_peaks_nb = nb.typed.List()
+    for i in bins_indices:
+        bins_peaks_nb.append(bins_peaks[i])
+
+    return bins_indices, bins_peaks_nb
+
+
+@nb.njit(cache=True)
+def _outlier_rejection(
+    bins_indices: List[int],
+    bins_peaks: nb.typed.List,
+    n_min: float,
+    n_max: float,
+) -> Tuple[nb.typed.List]:
+    """
+    Remove outliers from binned spectra using the sigma clipping algorithm
+    from https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/10.1002/pmic.202300234.
+
+    Parameters
+    ----------
+    bins_indices : List[int]
+        The indices of the non-empty bins.
+    bins_peaks : nb.typed.List
+        The intensities for each bin.
+    n_min : float
+        The number of standard deviations for the lower bound.
+    n_max : float
+        The number of standard deviations for the upper bound.
+
+    Returns
+    -------
+    nb.typed.List
+        The cleaned intensities for each bin.
+    """
+    cleaned_bins_peaks = nb.typed.List.empty_list(nb.types.float32[:])
+    for _ in range(len(bins_indices)):
+        cleaned_bins_peaks.append(np.empty(0, np.float32))
+
+    zero_peaks = np.zeros(len(bins_indices), dtype=np.bool_)
+    for i in range(len(bins_indices)):
+        intensities = bins_peaks[i]
+        if len(intensities) > 2:
+            clipped = _sigma_clipping(intensities, n_min, n_max)
+            if len(clipped) < 1:
+                zero_peaks[i] = True
+            else:
+                cleaned_bins_peaks[i] = clipped
+        else:
+            cleaned_bins_peaks[i] = intensities
+
+    # Remove empty bins
+    bins_indices = bins_indices[~zero_peaks]
+    new_cleaned_bins_peaks = []
+    for i in range(len(cleaned_bins_peaks)):
+        if not zero_peaks[i]:
+            new_cleaned_bins_peaks.append(cleaned_bins_peaks[i])
+
+    return bins_indices, new_cleaned_bins_peaks
+
+
+@nb.njit(cache=True)
+def _sigma_clipping(
+    intensities: np.ndarray, n_min: float, n_max: float
+) -> np.ndarray:
+    """
+    Apply sigma clipping to remove outliers from the array.
+
+    Parameters
+    ----------
+    intensities : np.ndarray
+        The array of intensities.
+    n_min : float
+        The number of standard deviations for the lower bound.
+    n_max : float
+        The number of standard deviations for the upper bound.
+
+    Returns
+    -------
+    np.ndarray
+        The array of intensities with outliers removed.
+    """
+    while len(intensities) > 2:
```

Comment:
**thought:** Interesting, I didn't know this was performed iteratively. Don't you progressively remove a bunch of peaks that way?